app-ng/src/handlers.lisp

```diff
@@ -0,0 +1,156 @@
+                  (loop :with mv := (gethash name (qvm::classical-memories qvm))
+                        :for idx :below (qvm::memory-view-length mv)
+                        :collect (qvm:memory-ref qvm name idx) :into mem
+                        :finally (push mem (gethash name results))))
+                 ;; Give only some things back.
+                 ((alexandria:proper-list-p indexes)
+                  (loop :for idx :in indexes
+                        :collect (qvm:memory-ref qvm name idx) :into mem
+                        :finally (push mem (gethash name results))))
+                 (t
+                  (error "Invalid address parameter for memory named ~S." name))))
+             addresses)
+    results))
+
+(define-rpc-handler (handle-create-qvm "create-qvm") ((simulation-method #'parse-simulation-method)
+                                                      (num-qubits #'parse-num-qubits))
+  "Create the requested persistent QVM.
+
+SIMULATION-METHOD is a STRING naming the desired simulation method (see *AVAILABLE-SIMULATION-METHODS*).
+
+NUM-QUBITS is a non-negative integer and represents the maximum number of qubits available on the QVM."
+  (encode-json (alexandria:plist-hash-table
+                (list "token" (allocate-persistent-qvm simulation-method num-qubits)))))
+
+(define-rpc-handler (handle-delete-qvm "delete-qvm") ((qvm-token #'parse-qvm-token))
+  "Delete a persistent QVM.
+
+QVM-TOKEN is a valid persistent QVM token such as one returned by the CREATE-QVM RPC call."
+  (delete-persistent-qvm qvm-token)
+  (encode-json (format nil "Deleted persistent QVM ~D" qvm-token)))
+
+(define-rpc-handler (handle-qvm-info "qvm-info") ((qvm-token #'parse-qvm-token))
+  "Return some basic bookkeeping info about the specified QVM.
+
+QVM-TOKEN is a valid persistent QVM token such as one returned by the CREATE-QVM RPC call."
+  (encode-json (persistent-qvm-info qvm-token)))
+
+(define-rpc-handler (handle-run-program "run-program")
+                    ((qvm-token #'parse-optional-qvm-token)
+                     (simulation-method #'parse-optional-simulation-method)
+                     (compiled-quil #'parse-quil-string)
+                     (addresses #'parse-addresses))
+  "Run the requested COMPILED-QUIL program, either on a persistent QVM or an emphemeral QVM using the given SIMULATION-METHOD.
+
+QVM-TOKEN is a valid persistent QVM token as returned by the CREATE-QVM RPC call.
+
+SIMULATION-METHOD is a STRING naming the desired simulation method (see *AVAILABLE-SIMULATION-METHODS*).
+
+COMPILED-QUIL is a STRING containing a valid Quil program.
+
+ADDRESSES is a HASH-TABLE where the keys are mem register names and the values are either T to indicate that we should return all indices for the corresponding register, or else a LIST of the desired indices for that register.
+
+The caller must provide either QVM-TOKEN or SIMULATION-METHOD, but not both."
+  (unless (alexandria:xor (null qvm-token) (null simulation-method))
```

Comment:
Suggestion: `(alexandria:xor qvm-token simulation-method)`

Reply:
I keep flip-flopping on whether to add explicit null checks or not...